internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -180,3 +183,166 @@ func (*topic) IsRetryable(error) bool {
+	queue    string // the AMQP queue name
+	consumer string // the client-generated name for this particular subscriber
+
+	mu   sync.Mutex
+	ch   *amqp.Channel // AMQP channel used for all communication.
+	delc <-chan amqp.Delivery
+}
+
+var nextConsumer int64 // atomic
+
+func newSubscription(conn *amqp.Connection, name string) (*subscription, error) {
+	s := &subscription{
+		conn:     conn,
+		queue:    name,
+		consumer: fmt.Sprintf("c%d", atomic.AddInt64(&nextConsumer, 1)),
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if err := s.establishChannel(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// Must be called with s.mu held.
+func (s *subscription) establishChannel() error {
+	// TODO(jba): support context.Context
+	if s.ch != nil {
+		// We already have a channel; nothing to do.
+		return nil
+	}
+	// Create a new channel.
+	ch, err := s.conn.Channel()
+	if err != nil {
+		return err
+	}
+	// Subscribe to messages from the queue.
+	s.delc, err = ch.Consume(s.queue, s.consumer,
+		false, // autoAck
+		false, // exclusive
+		false, // noLocal
+		wait,
+		nil) // args
+	if err != nil {
+		return err
+	}
+	s.ch = ch
+	return nil
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	// TODO(jba): Improve the timing characteristics of this implementation:
+	// If messages come at a rate just below the timeout, then we can take a long time to gather maxMessages.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if err := s.establishChannel(); err != nil {
+		return nil, err
+	}
+
+	// Get all the waiting messages, up to maxMessages.
+	var ms []*driver.Message
+loop:
+	for len(ms) < maxMessages {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+
+		case d, ok := <-s.delc:
+			if !ok { // channel closed
+				s.ch = nil // re-establish the channel
+				if len(ms) > 0 {
+					return ms, nil
+				}
+				// TODO(jba): get more information for a better message
+				return nil, errors.New("delivery channel closed")
+			}
+			ms = append(ms, toMessage(d))
+		default:
+			break loop
+		}
+	}
+	if len(ms) > 0 {
+		return ms, nil
+	}
+	// No messages; sleep for a bit so that the concrete implementation doesn't busy-wait.
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-time.After(50 * time.Millisecond):
+		return nil, nil
+	}
+}
+
+// toMessage converts an amqp.Delivery (a received message) to a driver.Message.
+func toMessage(d amqp.Delivery) *driver.Message {
+	// Delivery.Headers is a map[string]interface{}, so we have to
+	// convert each value to a string.
+	md := map[string]string{}
+	for k, v := range d.Headers {
+		md[k] = fmt.Sprint(v)
+	}
+	return &driver.Message{
+		Body:     d.Body,
+		AckID:    d.DeliveryTag,
+		Metadata: md,
+	}
+}
+
+// SendAcks implements driver.Subscription.SendAcks.
+func (s *subscription) SendAcks(ctx context.Context, ackIDs []driver.AckID) error {
+	s.mu.Lock()
```

Comment:
It seems unfortunate that `SendAcks` is blocked by `ReceiveBatch` due to the mutex. Maybe they should be on different channels? OK to file an issue to follow up on this separately.

Reply:
#853.